File: HPWorkpath_v1.6.3/Samples/ExampleAPIServices/source/StatisticsSample/src/com/hp/workpath/sample/statisticsample/fragment/StatisticListFragment.java
// Copyright 2018 HP Development Company, L.P.
// SPDX-License-Identifier: MIT
package com.hp.workpath.sample.statisticsample.fragment;

import static com.hp.workpath.sample.statisticsample.StatisticActivity.INDEX;
import static com.hp.workpath.sample.statisticsample.StatisticActivity.LAST_JOB;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.hp.workpath.api.Result;
import com.hp.workpath.api.statistics.StatisticsJobData;
import com.hp.workpath.sample.statisticsample.Logger;
import com.hp.workpath.sample.statisticsample.MainActivity;
import com.hp.workpath.sample.statisticsample.R;
import com.hp.workpath.sample.statisticsample.StatisticActivity;
import com.hp.workpath.sample.statisticsample.dialog.CommitDialogFragment;
import com.hp.workpath.sample.statisticsample.task.JobInfoTask;
import com.hp.workpath.sample.statisticsample.view.StatisticListAdapter;

import java.util.List;

public class StatisticListFragment extends Fragment implements View.OnClickListener {

    private static final int DIALOG_FRAGMENT = 1;
    private static final String SCREEN_4_3_INCH = "Screen_4.3_Inch";

    private Button mCommitButton;
    private Button mGetLastJobInfoButton;
    private RecyclerView mListView;
    private StatisticListAdapter mListAdapter;
    private ProgressBar mProgressBar;

    private List<StatisticsJobData> mJobDataList;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_statistic_list, container, false);
        mCommitButton = view.findViewById(R.id.commitButton);
        mGetLastJobInfoButton = view.findViewById(R.id.getLastJobInfoButton);
        mListView = view.findViewById(R.id.jobInfoListView);
        mListAdapter = new StatisticListAdapter(listOnClickListener);
        LinearLayoutManager layoutManager = new LinearLayoutManager(getContext());
        layoutManager.setReverseLayout(true);
        layoutManager.setStackFromEnd(true);
        mListView.setLayoutManager(layoutManager);
        mListView.setAdapter(mListAdapter);
        DividerItemDecoration dividerItemDecoration = new DividerItemDecoration(getContext(),
                DividerItemDecoration.VERTICAL);
        mListView.addItemDecoration(dividerItemDecoration);
        mProgressBar = view.findViewById(R.id.progressBar);
        mCommitButton.setOnClickListener(this);
        mGetLastJobInfoButton.setOnClickListener(this);
        if (SCREEN_4_3_INCH.equals((String)view.findViewById(R.id.container).getTag())) {
            view.findViewById(R.id.fabMenu).setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    if(view.findViewById(R.id.layoutBottom).getVisibility() == View.VISIBLE){
                        view.findViewById(R.id.layoutBottom).setVisibility(View.GONE);
                    }else{
                        view.findViewById(R.id.layoutBottom).setVisibility(View.VISIBLE);
                    }
                }
            });
        }
        return view;
    }

    @Override
    public void onResume() {
        super.onResume();
        ((MainActivity) getActivity()).refreshJobInfoList();
        if (mJobDataList != null) {
            mListAdapter.setItem(mJobDataList);
        }
    }

    View.OnClickListener listOnClickListener = new View.OnClickListener() {
        @Override
        public void onClick(View v) {
            int itemPosition = mListView.getChildLayoutPosition(v);
            View childView = mListView.getLayoutManager().findViewByPosition(itemPosition);
            TextView sequenceTextView = childView.findViewById(R.id.sequenceTextView);
            int jobSequence = Integer.parseInt(sequenceTextView.getText().toString());
            startJobInfoActivity(jobSequence);

        }
    };

    private void startLastJobInfoActivity() {
        if (mJobDataList != null && mJobDataList.size() > 0) {
            Intent intent = new Intent(getContext(), StatisticActivity.class);
            intent.putExtra(LAST_JOB, true);
            startActivity(intent);
        } else {
            if (getActivity() != null) {
                Logger.showResult(getActivity(), getString(R.string.no_job_info));
            }
        }
    }

    private void startJobInfoActivity(int index) {
        if (mJobDataList != null && index >= 0) {
            Intent intent = new Intent(getContext(), StatisticActivity.class);
            intent.putExtra(INDEX, index);
            startActivity(intent);
        } else {
            if (getActivity() != null) {
                Logger.showResult(getActivity(), getString(R.string.over_range));
            }
        }
    }

    public void getJobInfoList() {
        mProgressBar.setVisibility(View.VISIBLE);
        mListAdapter.clear();
        new JobInfoTask(getContext(), statisticsInterface).taskExecute();
    }

    @Override
    public void onClick(View v) {
        if (v == mCommitButton) {
            int totalTaskCount = ((MainActivity) getActivity()).getTotalTaskCount();
            CommitDialogFragment dialog = CommitDialogFragment.newInstance(totalTaskCount);
            dialog.setTargetFragment(this, DIALOG_FRAGMENT);
            dialog.show(getFragmentManager(), "dialog");
        } else if (v == mGetLastJobInfoButton) {
            startLastJobInfoActivity();
        }
    }

    ResponseInterface statisticsInterface = new ResponseInterface() {
        @Override
        public void success(List<StatisticsJobData> info) {
            mProgressBar.setVisibility(View.GONE);
            mJobDataList = info;
            if (mJobDataList != null && mJobDataList.size() > 0) {
                mListAdapter.setItem(mJobDataList);
            } else {
                if (getActivity() != null) {
                    Logger.showResult(getActivity(), getString(R.string.no_job_info));
                }
            }
        }

        @Override
        public void failure(String msg, Result result) {
            mProgressBar.setVisibility(View.GONE);
            Logger.showResult(getActivity(), msg, result);
        }
    };

    @Override
    public void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (resultCode == Activity.RESULT_OK) {
            switch (requestCode) {
                case DIALOG_FRAGMENT:
                    ((MainActivity) getActivity()).refreshJobInfoList();
                    break;
            }
        }
    }
}

File: HPWorkpath_v1.6.3/Samples_Kotlin/ExampleAPIServices/source/StatisticsSample/src/com/hp/workpath/sample/statisticsample/fragment/StatisticListFragment.kt
// Copyright 2018 HP Development Company, L.P.
// SPDX-License-Identifier: MIT
package com.hp.workpath.sample.statisticsample.fragment

import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.fragment.app.Fragment
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.hp.workpath.api.Result
import com.hp.workpath.api.statistics.StatisticsJobData
import com.hp.workpath.sample.statisticsample.Logger
import com.hp.workpath.sample.statisticsample.MainActivity
import com.hp.workpath.sample.statisticsample.R
import com.hp.workpath.sample.statisticsample.StatisticActivity
import com.hp.workpath.sample.statisticsample.StatisticActivity.Companion.REQUEST_KEY
import com.hp.workpath.sample.statisticsample.databinding.FragmentStatisticListBinding
import com.hp.workpath.sample.statisticsample.dialog.CommitDialogFragment
import com.hp.workpath.sample.statisticsample.task.JobInfoTask
import com.hp.workpath.sample.statisticsample.view.StatisticListAdapter
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch


class StatisticListFragment : Fragment(), View.OnClickListener {

    private lateinit var mListView: RecyclerView
    private var mListAdapter: StatisticListAdapter? = null
    private lateinit var mJobDataList: List<StatisticsJobData>
    private var mBindingFragment: FragmentStatisticListBinding? = null
    private val mBindingStatisticListFragment get() = mBindingFragment!!
    private val SCREEN_4_3_INCH = "Screen_4.3_Inch"
    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        mBindingFragment =
            FragmentStatisticListBinding.inflate(inflater, container, false)
        return mBindingStatisticListFragment.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        mListView = view.findViewById(R.id.jobInfoListView)
        mListAdapter = StatisticListAdapter(listOnClickListener)
        val layoutManager = LinearLayoutManager(context)
        layoutManager.reverseLayout = true
        layoutManager.stackFromEnd = true
        mListView.layoutManager = layoutManager
        mListView.adapter = mListAdapter
        val dividerItemDecoration = DividerItemDecoration(
            context,
            DividerItemDecoration.VERTICAL
        )
        mListView.addItemDecoration(dividerItemDecoration)
        mBindingStatisticListFragment.commitButton.setOnClickListener(this)
        mBindingStatisticListFragment.getLastJobInfoButton.setOnClickListener(this)

        (activity as MainActivity?)?.refreshJobInfoList()
        if (::mJobDataList.isInitialized) {
            mListAdapter?.setItem(mJobDataList as MutableList<StatisticsJobData>)
        }
        if (SCREEN_4_3_INCH.equals(mBindingStatisticListFragment.container.tag)) {
            mBindingStatisticListFragment.fabMenu?.setOnClickListener {
                if (mBindingStatisticListFragment.layoutBottom.visibility == View.VISIBLE) {
                    mBindingStatisticListFragment.layoutBottom.visibility = View.GONE
                } else {
                    mBindingStatisticListFragment.layoutBottom.visibility = View.VISIBLE
                }
            }
        }
    }

    private var listOnClickListener = View.OnClickListener { v ->
        val itemPosition = mListView.getChildLayoutPosition(v)
        val childView = mListView.layoutManager?.findViewByPosition(itemPosition)
        val sequenceTextView = childView?.findViewById<TextView>(R.id.sequenceTextView)
        val jobSequence = sequenceTextView?.text.toString().toInt()
        startJobInfoActivity(jobSequence)
    }

    private fun startLastJobInfoActivity() {
        if (::mJobDataList.isInitialized) {
            val intent = Intent(context, StatisticActivity::class.java)
            intent.putExtra(StatisticActivity.LAST_JOB, true)
            startActivity(intent)
        } else {
            if (activity != null) {
                Logger.showResult(activity, getString(R.string.no_job_info))
            }
        }
    }

    private fun startJobInfoActivity(index: Int) {
        if (index >= 0) {
            val intent = Intent(context, StatisticActivity::class.java)
            intent.putExtra(StatisticActivity.INDEX, index)
            startActivity(intent)
        } else {
            if (activity != null) {
                Logger.showResult(activity, getString(R.string.over_range))
            }
        }
    }

    fun getJobInfoList() {
        mBindingStatisticListFragment.progressBar.visibility = View.VISIBLE
        mListAdapter?.clear()
        viewLifecycleOwner.lifecycleScope.launch(Dispatchers.Default) {
            JobInfoTask(requireContext(), statisticsInterface).execute()
        }
    }

    override fun onClick(v: View) {
        if (v === mBindingStatisticListFragment.commitButton) {
            val dialog = CommitDialogFragment.newInstance()
            dialog.show(parentFragmentManager, "dialog")
            parentFragmentManager.setFragmentResultListener(
                REQUEST_KEY,
                this
            ) { requestKey, _ ->
                if (REQUEST_KEY == requestKey) {
                    (activity as MainActivity).refreshJobInfoList()
                }
            }
        } else if (v === mBindingStatisticListFragment.getLastJobInfoButton) {
            startLastJobInfoActivity()
        }
    }

    private var statisticsInterface: ResponseInterface = object : ResponseInterface {
        override fun success(info: List<StatisticsJobData?>) {
            mBindingStatisticListFragment.progressBar.visibility = View.GONE

            mJobDataList = info.filterNotNull()
            if (mJobDataList.isNotEmpty()) {
                mListAdapter?.setItem(mJobDataList.toMutableList())
            } else {
                if (activity != null) {
                    Logger.showResult(activity, getString(R.string.no_job_info))
                }
            }
        }

        override fun failure(msg: String?, result: Result?) {
            mBindingStatisticListFragment.progressBar.visibility = View.GONE
            Logger.showResult(activity, msg, result)
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        mBindingFragment = null
    }
}